File: Algorithms/Sorting/Merge Sort/merge-sort.go
/*
 * Merge sort
 *
 */

package main

import (
	"fmt"
	"math/rand"
)

func randomArray(n int) []int {
	arr := make([]int, n)
	for i := 0; i < n; i++ {
		arr[i] = rand.Intn(n)
	}
	return arr
}

func Merge(l, r []int) []int {
	result := make([]int, 0, len(l)+len(r))

	for len(l) > 0 || len(r) > 0 {
		if len(l) == 0 {
			return append(result, r...)
		}
		if len(r) == 0 {
			return append(result, l...)
		}
		if l[0] <= r[0] {
			result = append(result, l[0])
			l = l[1:]
		} else {
			result = append(result, r[0])
			r = r[1:]
		}
	}

	return result
}

func MergeSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	middle := len(arr) / 2

	l := MergeSort(arr[:middle])
	r := MergeSort(arr[middle:])

	return Merge(l, r)
}

func main() {
	arr := randomArray(10)
	fmt.Println("Initial array is:", arr)
	fmt.Println("")
	fmt.Println("Sorted array:    ", MergeSort(arr))
}

File: Algorithms/Sorting/Merge Sort/merge-sort.js
/*
 * Merge sort
 *
*/


function Merge(arr, left, right, middle){
    let n1 = middle - left + 1,
        n2 = right - middle,
        L = [],
        R = []

    for(let i =0;i<n1;i++)
        L[i] = arr[left + i]

    for(let j=0;j<n2;j++)
        R[j]=arr[middle+j+1]

    i = 0
    j = 0
    let k = left
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i]
            i++
        } else {
            arr[k] = R[j]
            j++
        }
        k++
    }

    while (i < n1) {
        arr[k] = L[i]
        i++
        k++
    }

   while (j < n2) {
        arr[k] = R[j]
        j++
        k++
    }
}

function MergeSort(arr, left, right) {
    if(left < right) {
        let middle = Math.floor((left +right)/2)

        MergeSort(arr, left, middle)
        MergeSort(arr, middle+1, right)
        Merge(arr, left, right, middle)
    }
}


arr = [1,3,6,54,56,78,78,6754,643,5,4]
MergeSort(arr, 0, arr.length-1)
console.log(arr)
